python/tvm/driver/tvmc/byoc.py

```diff
@@ -0,0 +1,191 @@
+"""
+import logging
+
+from abc import ABC
+from abc import abstractmethod
+
+import tvm
+
+from tvm import relay
+
+from tvm.relay.op.contrib import get_pattern_table
+
+from .common import TVMCException
+
+
+# pylint: disable=invalid-name
+logger = logging.getLogger("TVMC")
+
+# Global dictionary to map existing custom codegens
+# with the names used to use them
+REGISTERED_CODEGEN = {}
+
+
+def register_codegen(kind):
+    """
+    Utility function to register a BYOC class for TVMC.
+
+    Classes decorated with `tvm.driver.tvmc.target.register_codegen` will
+    be added to the codegens dictionary.
+
+    Example
+    -------
+
+        @register_codegen(kind="samplebyoc")
+        class MyCustomTarget(TVMCCodegen):
+            ...
+    """
+
+    def codegen_decorator(cls):
+        cls.kind = kind
+
+        assert kind not in REGISTERED_CODEGEN, "there is already a codegen '%s': %s" % (
+            kind,
+            REGISTERED_CODEGEN[kind],
+        )
+        assert issubclass(cls, TVMCCodegen), "%s is expected to be a subclass of TVMCCodegen" % cls
+
+        REGISTERED_CODEGEN[kind] = cls
+        return cls
+
+    return codegen_decorator
+
+
+def get_codegen_kinds():
+    """Return a list of all registered codegens.
+
+    Returns
+    --------
+    list of str
+        all registered codegens
+    """
+    return REGISTERED_CODEGEN.keys()
+
+
+def get_codegen_by_kind(kind):
+    """Return a custom codegen by kind.
+
+    Returns
+    --------
+    TVMCCodegen
+        The requested codegen or None in case it is
+        not valid
+    """
+    try:
+        return REGISTERED_CODEGEN[kind]
+    except KeyError:
+        raise TVMCException("Target %s is not defined." % kind)
+
+
+class TVMCCodegen(ABC):
+    """Abstract class for command line driver BYOC definition.
+
+    Provide a unified way to create a codegen adapter with a set of
+    callback functions to be used within TVMC.
+
+    """
+
+    @staticmethod
+    @abstractmethod
+    def get_config_key_name():
+        """Return the name of the dictionary key to be used
+        at compile time (i.e. relay.build) as part of the
+        "config" argument.
+
+        Returns
+        -------
+        str
+            Name of the Codegen config dictionary key
+            to match with the expected name in the relay
+            implementation of this Codegen.
+        """
+
+    @staticmethod
+    @abstractmethod
+    def run_custom_passes(mod, params):
+        """Apply a set of transformations to the module
+        before compilation happens.
+
+        Parameters
+        ----------
+        mod : tvm.relay.Module
+            The relay module to convert.
+        params : dict
+            The parameters (weights) for the TVM module.
+
+        Returns
+        -------
+        mod: tvm.relay.Module
+            The converted module.
+        """
+
+
+@register_codegen(kind="ethos-n77")
+class TVMCEthosNCodegen(TVMCCodegen):
```

Comment:
Once we have the feature, I think it would be better to have them added as simpler self-contained patches. What do you think?

Reply:
If adding a codegen is just adding a class like this, it seems to me that it is not necessary to have separate PRs. After all, we are not testing the codegen flow in this module.

Reply:
I also added `partition_for_arm_compute_lib` in the context of this PR.

To add TensorRT, I'd prefer adding it on a different patch, mostly because I has a different method signature, and we'll need to see how to accommodate those, or even whether we should have an _agreed_ function signature for the `partition_to_*` functions. I think we should, but I'd like to discuss that in a separate PR.

Reply:
I'll leave this open, in case you have any concerns, or to close in case you agree with my suggestion above.

Reply:
In fact, this is exactly the reason I would like to add others -- make sure this registration mechanism works for not only one or certain codegens. If adding TensorRT would change the way we register the codegen, the registry should be improved. Of course, it's also possible to keep the current registry map and change the TensorRT partition function. It would be better to at least figure out what to change in this PR.

cc @trevor-m

Reply:
So in the specific case, we can see that Compute Library and Arm Ethos-N use the same, similar to passes signature, taking only mod and params, whereas TensorRT takes some other optional arguments.

If  can just call it the way we do for the others, all the others will be used as defaults. If we can do that, I'm happy to add it to the patch.

Compute Library:
https://github.com/apache/tvm/blob/fc48514f1d8ccffcebd12007cb6c602506975703/python/tvm/relay/op/contrib/arm_compute_lib.py#L46

Arm Ethos-N:
https://github.com/apache/tvm/blob/e1ce6234971a5bae0f81d7dc0e947ae9d2e8436b/python/tvm/relay/op/contrib/ethosn.py#L49

TensorRT:
https://github.com/apache/tvm/blob/fc48514f1d8ccffcebd12007cb6c602506975703/python/tvm/relay/op/contrib/tensorrt.py#L81-L88

I think that will be a challenge for the partition as whole, if we don't standardise the way to have the `partition_to_*` functions. What do you think?